lib/improver/tests/nbhood/circular_kernel/test_CircularPercentiles.py

```diff
@@ -0,0 +1,425 @@
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for the nbhood.circular_kernel.CircularPercentiles plugin."""
+
+
+import unittest
+
+import iris
+from iris.cube import Cube
+from iris.tests import IrisTest
+import numpy as np
+
+from improver.nbhood.circular_kernel import CircularPercentiles
+from improver.tests.nbhood.nbhood.test_NeighbourhoodProcessing import (
+    set_up_cube, set_up_cube_lat_long)
+from improver.percentile import PercentileConverter
+
+# Standard percentile answers for radius 2 grid points
+# (==> circle of 13 points)
+# Array name indicates how many of the 13 grid points were ones;
+# remaining points are zeroes.
+PERCENTILES_1_IN_13 = np.ones(15)
+PERCENTILES_1_IN_13[:2] = [0., 0.6]
+PERCENTILES_2_IN_13 = np.ones(15)
+PERCENTILES_2_IN_13[:3] = [0., 0., 0.2]
+
+# For 1 zero in group of 5 (other 4 are ones):
+PERCENTILES_1_IN_5 = np.ones(15)
+PERCENTILES_1_IN_5[:4] = [0., 0.2, 0.4, 0.8]
+
+# Standard percentile answers for radius 3 grid points
+# (==> circle of 25 points)
+PERCENTILES_1_IN_25 = np.ones(15)
+PERCENTILES_1_IN_25[:1] = [0.]
+
+
+class Test__repr__(IrisTest):
+
+    """Test the repr method."""
+
+    def test_basic(self):
+        """Test that the __repr__ returns the expected string."""
+        result = str(CircularPercentiles())
+        msg = '<CircularPercentiles: percentiles: {}>'.format(
+            PercentileConverter.DEFAULT_PERCENTILES)
+        self.assertEqual(str(result), msg)
+
+
+class Test_make_percentile_cube(IrisTest):
+
+    """Test CircularPercentiles.make_percentile_cube."""
+
+    def test_basic(self):
+        """
+        Test that the plugin returns an iris.cube.Cube and that percentile
+        coord is added.
+        """
+        cube = set_up_cube(
+            zero_point_indices=((0, 0, 2, 2),), num_time_points=1,
+            num_grid_points=5)
+        result = (
+            CircularPercentiles().make_percentile_cube(cube))
+        self.assertIsInstance(result, Cube)
+        self.assertIsInstance(result.coord(
+            'percentiles_over_neighbourhood'), iris.coords.Coord)
+        self.assertArrayEqual(result.coord(
+            'percentiles_over_neighbourhood').points,
+            PercentileConverter.DEFAULT_PERCENTILES)
+
+
+class Test_run(IrisTest):
+
+    """Test neighbourhood percentile plugin on the OS National Grid."""
+
+    def test_basic(self):
+        """Test that the plugin returns an iris.cube.Cube."""
+        cube = set_up_cube(
+            zero_point_indices=((0, 0, 2, 2),), num_time_points=1,
+            num_grid_points=5)
+        ranges = 2
+        result = CircularPercentiles().run(cube, ranges)
+        self.assertIsInstance(result, Cube)
+
+    def test_single_point(self):
+        """Test behaviour for a single non-zero grid cell."""
+        cube = set_up_cube()
+        expected = np.ones([15] + list(np.shape(cube.data)))
```

Comment:
Generally these are a good set of unit tests. I think we can do these tests on a smaller grid, which may make it easier to see what is going one and avoid creating expected results by tiling etc. It would be good to test on a case where the circle is bigger than the grid.